onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -0,0 +1,283 @@
+            return check_result.fail(
+                f"Shape mismatch: {query_mm_sliced} does not match expected dimensions ['B', 'S', 'Dh_q']",
+                query_mm_sliced,
+            )
+        if no_match(key_mm_sliced, ["B", "S", "Dh_k"]):
+            return check_result.fail(
+                f"Shape mismatch: {key_mm_sliced} does not match expected dimensions ['B', 'S', 'Dh_k']",
+                key_mm_sliced,
+            )
+        if no_match(value_mm_sliced, ["B", "S", "Dh_v"]):
+            return check_result.fail(
+                f"Shape mismatch: {value_mm_sliced} does not match expected dimensions ['B', 'S', 'Dh_v']",
+                value_mm_sliced,
+            )
+
+        # Ensure Dh = Dh_q + Dh_k + Dh_v
+        Dh = bindings.get("Dh")
+        Dh_q = bindings.get("Dh_q")
+        Dh_k = bindings.get("Dh_k")
+        Dh_v = bindings.get("Dh_v")
+
+        if (
+            not isinstance(Dh, int)
+            or not isinstance(Dh_q, int)
+            or not isinstance(Dh_k, int)
+            or not isinstance(Dh_v, int)
+        ):
+            return False  # Missing bindings, cannot verify
+
+        if Dh != Dh_q + Dh_k + Dh_v:  # type: ignore[operator]
+            return check_result.fail(
+                f"Hidden size of query, key and value do not add up to hidden size: {Dh} != {Dh_q} + {Dh_k} + {Dh_v}",
+            )
+
+        # TODO: Add mask check once mask is added to the pattern
+        return check_result
+
+    def rewrite(
+        self,
+        op,
+        input,
+        qkv_weight,
+        qkv_bias,
+        # mask_index,
+        past,
+        # attention_bias,
+        num_heads,
+        # scale,
+        **_,
+    ):
+        if self._has_past:
+            attention, present = op.Attention(
+                input,
+                qkv_weight,
+                qkv_bias,
+                None,  # mask_index
+                past,
+                # attention_bias,
+                # past_sequence_length
+                num_heads=num_heads,
+                # qkv_hidden_sizes=qkv_hidden_sizes,
+                # scale=scale,
+                _domain="com.microsoft",
+                _outputs=2,
+            )
+            # TODO: Switch back order of outputs
+            return present, attention
+        else:
+            return op.Attention(
+                input,
+                qkv_weight,
+                qkv_bias,
+                # mask_index
+                # past
+                # attention_bias,
+                # past_sequence_length
+                num_heads=num_heads,
+                # qkv_hidden_sizes=qkv_hidden_sizes,
+                # scale=scale,
+                _domain="com.microsoft",
+                _outputs=1,
+            )
+
+
+attention_with_input_bias_rule = AttentionFusion.rule(
+    "attention_input_bias",
+    has_input_bias=True,
+    has_past=False,
+)
+attention_with_no_input_bias_rule = AttentionFusion.rule(
```

Comment:
Just a minor suggestion for the future (this looks fine for now): I found that when the number of combinations increases, it is easier to use names listing only positive flags: that is, names like "attention_with_bias" and "attention" (which implicitly has no bias).  Avoids very long names if you have 3 flags, for example. In this case, for example, "attention", "attention_with_bias", "attention_with_past", "attention_with_bias_and_past".

Reply:
Modified